tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -11299,6 +11299,62 @@ bool Sema::DiagnoseHLSLMethodCall(const CXXMethodDecl *MD, SourceLocation Loc) {
   return false;
 }
 
+void ValidateWaveSize(clang::Sema *S, FunctionDecl *FD) {
+  const auto *SM =
+      hlsl::ShaderModel::GetByName(S->getLangOpts().HLSLProfile.c_str());
+
+  const llvm::StringRef &entryName = S->getLangOpts().HLSLEntryFunction;
+  StringRef functionName = "";
+  if (FD->getIdentifier())
+    functionName = FD->getName();
+  bool isEntry = false;
+  bool isNode = false;
+  bool isCS = false;
+
+  isEntry =
+      !SM->IsLib() &&
+      FD->getDeclContext()->getDeclKind() == Decl::Kind::TranslationUnit &&
+      functionName == entryName;
```

Comment:
There needs to be some code somewhere that is responsible for emitting a diagnostic when, say, the wavesize attribute is found on, say, a pixel shader. That's why I think the isNode and isCS bools are necessary. I agree that isEntry seems useless since this function is only called on entry point decls anyways.

Agree with your last statement about the extra exemption, I'll add that.

Reply:
> As for your last point, I believe my current implementation satisfies your description, since it doesn't bother validating the wavesize attr on non-entry point decl's at all.

This is because the function is only called when this is the active entry point in a non-library compilation.  That's not correct, since we want to diagnose all known entry points (ones marked with the shader attribute).  However _if we do that_, diagnostics that depend on compilation options should not occur for entry points that aren't active, since those options could be different for a compilation where those entry points _are_ active.

Does that make sense?